### DISCUSSION TITLE
Adding Cementite to existing database

### DISCUSSION THREAD

**Discussion by andr1976** [reaktoro/reaktoro#293] [Q&A]
### Description

What is the easiest way of adding cementite/Fe3C to an existing database. I have not been able to find it in any of the available databses in reaktoro? 

### What is your operating system?

None

**Comment by allanleal** [comment]
Hi Anders, I'm leaving on vacation now. I'll get back to you with an idea next week. But for anything cement, @gdmiron should be able to help with ThermoFun + Reaktoro.

**Comment by gdmiron** [comment]
Hello @andr1976, while this phase is more related to metallurgy than to cement it is possible to add it to a database used by thermofun and then use thermofun to initialize a reaktoro system. You first need a main database that covers most of your system and the you can have a user defined file where you keep additional substances to append to the main database. The easiest way is to copy one existing record for a solid in the database and replace the fields with data for Fe3C (reference thermodynamic properties, Cp temperature coefficients).

Here there is an example of appending user data to an existing database:
https://mybinder.org/v2/gh/thermohub/thermofun-jupyter/master?urlpath=lab/tree/how-to-use-a-database.ipynb (cell 9)

In short:
```python
import thermofun as fun

fundb = fun.Database('path to a thermofun database') # this would be the main database

fundb.appendData('mydata-thermofun.json')

```

after you have everything you need in the thermofun db object, `fundb`, you do
```python

# Define the Thermofun database - here thermofun fundb is interfaced with reaktoro
db = ThermoFunDatabase(fundb)

# Define reaktoro phases eg. aq phase
solution = AqueousPhase(speciate("C Cl H O P S U"))
solution.set(chain(
    ActivityModelHKF(),
    ActivityModelDrummond("CO2")
))

# Define chemical system by providing a database and an aqueous phase
system = ChemicalSystem(db, solution)
```

You may also look here: https://reaktoro.org/tutorials/advanced/creating-custom-databases.html

**Comment by allanleal** [comment]
Hi Anders, the file below is the modified NASA-CEA database in Reaktoro YAML format with the first species representing the added cementite species. I had to zip it to make it available here:

[nasa-cea-with-cementite.zip](https://github.com/reaktoro/reaktoro/files/11163903/nasa-cea-with-cementite.zip)

The reason for picking the NASA-CEA database here is that some Reaktoro users have successfully used it for modeling some metallurgical processes.

The text block added to the original `nasa-cea.yaml` database is:

![image](https://user-images.githubusercontent.com/5825588/230231487-837350cd-10ea-4607-b645-6800d18befc0.png)

The standard Gibbs energy of formation at 673 K and 723 K were obtained from [Lee and Simkovich (1988)](https://link.springer.com/article/10.1007/BF02645214). More exactly, by evaluating the following linear equation at those two temperatures:

![image](https://user-images.githubusercontent.com/5825588/230231870-54243203-cb5e-416a-9180-fe449aa2f9f0.png)

In case you have other source of thermodynamic properties for cementite (including perhaps its standard molar enthalpy of formation, standard molar volume, etc.), for wider range of temperatures, we can accommodate this via interpolation or by implementing the model directly in Python or C++.

I have not tested the above modified database within an equilibrium calculation. I also simply checked the paper quickly and found the expression above. More due diligence may be needed. I just wanted to illustrate here a possibility to add a new species with data available in the literature. Let us know how you progress from here and if you encounter new needs or issues.

**Reply by andr1976** [reply]
Thanks to both of you @allanleal  and @gdmiron, thanks a lot. This should get me started.
